File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    static ArrayList<Integer>[] graph;
    static int min;
    
    public int solution(int n, int[][] wires) {
        graph = new ArrayList[n+1];
        min = Integer.MAX_VALUE;
        
        //그래프 ArrayList 초기화, 노드 개수만큼 생성
        for(int i=0; i<=n; i++){
            graph[i] = new ArrayList<>();
        }
        
        //양방향 간선 구조
        for(int i=0; i<wires.length; i++){
            int v1 = wires[i][0];
            int v2 = wires[i][1];
            graph[v1].add(v2);
            graph[v2].add(v1);
        }
        
        //여기서부터 탐색 시작
        for(int i=0; i<wires.length; i++){
            int v1 = wires[i][0];
            int v2 = wires[i][1];
            
            boolean[] visited = new boolean[n+1];
            
            //해당 간선을 그래프에서 제거
            graph[v1].remove(Integer.valueOf(v2));
            graph[v2].remove(Integer.valueOf(v1));
            
            int cnt = dfs(1, visited); //임의의 시작점에서 dfs 탐색
            
            int diff = Math.abs(cnt - (n-cnt));
            min = Math.min(diff, min);
            
            //그래프에 다시 간선 추가
            graph[v1].add(v2);
            graph[v2].add(v1);
        }
        return min;
    }
    
    static int dfs(int v, boolean[] visited){
        visited[v] = true;
        int cnt = 1;
        
        for(int next: graph[v]){
            if(!visited[next]){
                cnt += dfs(next, visited);
            }
        }
        return cnt;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    answer = float('inf')
    graph = [[] for _ in range(n+1)]
    # 송전탑 연결 트리 만들기
    for wire in wires:
        a,b = wire
        graph[a].append(b)
        graph[b].append(a)
    
    
    def dfs(node, cut_a, cut_b, visited):
        visited[node] = True
        count = 1
        
        for next_node in graph[node]:
            # 끊은 전선이면 이동하지 않음
            if node == cut_a and next_node == cut_b:
                continue
                
            if node == cut_b and next_node == cut_a:
                continue
                
            if not visited[next_node]:
                count += dfs(next_node, cut_a, cut_b, visited)
        return count
    
    
    for cut_a, cut_b in wires:
        # a,b 전선 끊었다고 가정
        visited = [False] * (n+1)
        
        #a에서 DFS 시작
        # 단 a-b 연결은 못 지나가게 처리
        count = dfs(cut_a, cut_a, cut_b, visited)
        other_count = n - count
        
        # 두 전력망의 송전탑 개수 차이
        diff = abs(count - other_count)
        answer = min(answer, diff)
    return answer